library/L2_HAL/sensors/battery/ltc4150.hpp

```diff
@@ -4,95 +4,74 @@
   };
-
-  /// Class that handles the ticks from the LTC4150. The class uses a template
-  /// so there can be different handlers for different LTC4150s connected to a
-  /// board.
-  template <int id>
-  class TickHandler
-  {
-   public:
-    /// This member variable will hold the number of ticks sent from the
-    /// LTC4150. The interrupt handler increments this value by one every time
-    /// it is invoked.
-    inline static Battery_t battery_info;
-
-    /// handles the ISR by incrementing the num_ticks variable of TickHandler's
-    /// Battery_t
-    static void IsrHandler()
-    {
-      battery_info.num_ticks += 1;
-    }
-
-    /// @return the calculated milliamp hours based on the number of ticks.
-    /// 3600 is X 32.55 is kGvF and 0.00005 is kRsense
-    static float GetBatterymAh()
-    {
-      return static_cast<float>(static_cast<float>(battery_info.num_ticks) /
-                                (3600 * kGvf * kRsense));
-    }
-  };
-
-  /// @param isr - handler for coulomb counter ticks.
   /// @param int_pin - gpio pin for handling handling interrupts from the
-  /// LTC4150
+  ///        LTC4150
   /// @param pol - gpio pin to determine the polarity output of the LTC4150.
-  template <int id>
-  explicit constexpr Ltc4150(const TickHandler<id> & isr,
-                             Gpio & int_pin,
-                             Gpio & pol)
+  /// @param resistance - value of impedance represented in ohms for
+  ///        calculating battery charge
+  explicit constexpr Ltc4150(sjsu::Gpio & int_pin,
+                             sjsu::Gpio & pol,
+                             units::impedance::ohm_t resistance)
       : int_pin_(int_pin),
-        pol_(pol),
-        get_mah_(isr.GetBatterymAh),
-        isr_handler_(isr.IsrHandler)
+        pol_pin_(pol),
+        resistance_(resistance),
+        pulses_(0),
+        polarity_(Polarity::kDischarging)
   {
   }
 
   /// Initialize hardware, setting pins as inputs and attaching ISR handlers to
   /// the interrupt pin.
-  void Initialize() const override
+  void Initialize() override
   {
-    pol_.SetAsInput();
+    pol_pin_.SetAsInput();
     int_pin_.SetAsInput();
-    int_pin_.AttachInterrupt(isr_handler_, Gpio::Edge::kEdgeFalling);
+    pol_pin_.AttachInterrupt(
+        [this]() {
+          polarity_ =
+              (pol_pin_.Read()) ? Polarity::kDischarging : Polarity::kCharging;
+        },
+        Gpio::Edge::kEdgeFalling);
+    int_pin_.AttachInterrupt([this]() { pulses_++; }, Gpio::Edge::kEdgeFalling);
   }
 
-  /// @return the calculated mAh from the TickHandler instance
-  float GetBatterymAh() const override
+  /// @return the calculated mAh
+  units::charge::milliampere_hour_t GetCharge() const override
   {
-    return get_mah_();
+    /// 3600 is X 32.55 is kGvF and 0.00005 is kRsense
+    auto sign = (polarity_ == Polarity::kCharging)
+                    ? units::dimensionless::scalar_t(-1)
+                    : units::dimensionless::scalar_t(1);
```

Comment:
See comment about about subtracting and adding to pulses based on the polarity. This doesn't make sense here as the pin could change at anytime during usage. If it was unchanging, we should could simply have the developer give us  a  bool parameter in the constructor to figure out which state it will be in rather than using a Gpio.

Reply:
I added an initial state parameter that was of type Ltc4150::Polarity instead of bool. Would this work? I also got rid of this line since the sign is accounted for in the updated polarity ISR.

Reply:
I removed the supplied parameter, and I read the gpio at `Initialize()` [here](https://github.com/kammce/SJSU-Dev2/blob/8b224131585cd1212a5a0c085fdf5660e2f82f8c/library/L2_HAL/sensors/battery/ltc4150.hpp#L61). To incorporate the `Initialize()` read in my tests, I added a change in mock behavor, for example with the cancel and change in sign test cases [here](https://github.com/kammce/SJSU-Dev2/blob/8b224131585cd1212a5a0c085fdf5660e2f82f8c/library/L2_HAL/sensors/battery/test/ltc4150_test.cpp#L155)